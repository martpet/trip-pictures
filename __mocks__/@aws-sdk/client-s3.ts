export const S3Client = function DummyS3Client() {};
