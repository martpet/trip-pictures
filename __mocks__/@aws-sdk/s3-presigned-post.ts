export const createPresignedPost = vi
  .fn()
  .mockName('createPresignedPost')
  .mockResolvedValue('dummyPresignedPost');
