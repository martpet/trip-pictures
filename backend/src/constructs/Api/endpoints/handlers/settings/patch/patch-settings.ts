import {
  APIGatewayProxyHandlerV2,
  createDynamoUpdateExpression,
  DynamoDBClient,
  DynamoDBDocumentClient,
  errorResponse,
  getIdTokenPayload,
  PatchSettingsRequest,
  PatchSettingsResponse,
  RouteHeaders,
  StatusCodes,
  SyncedSettingsKey,
  syncedSettingsKeys,
  UpdateCommand,
  UsersTableItem,
  usersTableOptions,
} from 'lambda-layer';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);
const settingsAttrName: keyof UsersTableItem = 'settings';
const allowedSettings = syncedSettingsKeys;

export const handler: APIGatewayProxyHandlerV2<PatchSettingsResponse> = async (event) => {
  const { authorization } = event.headers as RouteHeaders<'/settings'>;

  if (!authorization) {
    return errorResponse('UMxOJy1cpJ');
  }

  if (!event.body) {
    return errorResponse('x1IsNHbqd3', { statusCode: StatusCodes.BAD_REQUEST });
  }

  let patch;

  try {
    patch = JSON.parse(event.body) as PatchSettingsRequest;
  } catch (error) {
    return errorResponse('d6QTt6tKWK', { statusCode: StatusCodes.BAD_REQUEST, error });
  }

  const hasUnknownKeys = Object.keys(patch).some(
    (key) => !allowedSettings.includes(key as SyncedSettingsKey)
  );

  if (hasUnknownKeys) {
    return errorResponse('IUKROTi0UF', { statusCode: StatusCodes.BAD_REQUEST });
  }

  const { sub } = await getIdTokenPayload(authorization);

  const updateCommand = new UpdateCommand({
    TableName: usersTableOptions.tableName,
    Key: { [usersTableOptions.partitionKey.name]: sub },
    ...createDynamoUpdateExpression(patch, settingsAttrName),
  });

  await ddbDocClient.send(updateCommand);

  return {
    statusCode: StatusCodes.NO_CONTENT,
  };
};
