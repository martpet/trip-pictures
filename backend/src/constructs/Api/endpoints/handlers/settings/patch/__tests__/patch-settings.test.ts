import { DynamoDBDocumentClient, UpdateCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';

import {
  itGetsIdToken,
  itHasJsonBody,
  itResolves,
  itResolvesWithError,
  itSendsAwsCommand,
} from '~/constructs/Api/utils';

import { handler } from '../patch-settings';

vi.mock('~/constructs/Api/utils/getIdTokenPayload');
vi.mock('~/constructs/Api/utils/errorResponse');

const ddbMock = mockClient(DynamoDBDocumentClient);

const args = [
  {
    headers: { authorization: 'dummyAuthorizationHeader' },
    body: JSON.stringify({ language: 'dummyLanguage' }),
  },
] as unknown as Parameters<typeof handler>;

beforeEach(() => {
  ddbMock.reset();
  ddbMock.on(UpdateCommand).resolves({});
});

describe('patch-settings', () => {
  itHasJsonBody(handler, args);
  itGetsIdToken(handler, args);
  itSendsAwsCommand(UpdateCommand, ddbMock, handler, args);
  itResolves(handler, args);

  describe('when "event.body" has unknown keys', () => {
    const argsClone = structuredClone(args);
    argsClone[0].body = JSON.stringify({ dummyUnknownKey: '' });
    itResolvesWithError(handler, argsClone);
  });
});
