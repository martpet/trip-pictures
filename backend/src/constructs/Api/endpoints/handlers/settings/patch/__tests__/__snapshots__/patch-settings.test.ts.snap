// Vitest Snapshot v1

exports[`patch-settings > calls "getIdTokenPayload" with correct args 1`] = `
[
  [
    "dummyAuthorizationHeader",
  ],
]
`;

exports[`patch-settings > resolves with a correct value 1`] = `
{
  "statusCode": 204,
}
`;

exports[`patch-settings > sends "UpdateCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "ExpressionAttributeNames": {
    "#language": "language",
    "#settings": "settings",
  },
  "ExpressionAttributeValues": {
    ":language": "dummyLanguage",
  },
  "Key": {
    "id": "dummySub",
  },
  "TableName": "TP-users",
  "UpdateExpression": "set #settings.#language = :language",
}
`;

exports[`patch-settings > when "authorization" header is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "UMxOJy1cpJ",
  ],
]
`;

exports[`patch-settings > when "authorization" header is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`patch-settings > when "event.body" has unknown keys > calls "errorResponse" with correct args 1`] = `
[
  [
    "IUKROTi0UF",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`patch-settings > when "event.body" has unknown keys > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`patch-settings > when "event.body" is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "x1IsNHbqd3",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`patch-settings > when "event.body" is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`patch-settings > when "event.body" is not JSON > calls "errorResponse" with correct args 1`] = `
[
  [
    "d6QTt6tKWK",
    {
      "error": [SyntaxError: Unexpected token o in JSON at position 1],
      "statusCode": 400,
    },
  ],
]
`;

exports[`patch-settings > when "event.body" is not JSON > resolves with a correct value 1`] = `"dummyErrorResponse"`;
